### DISCUSSION TITLE
imagemagick VS2015 compile error.

### DISCUSSION THREAD

**Discussion by papafrog77** [ImageMagick/ImageMagick#5765] [Development] [answered]
i  m new in imagemagick. i follow instruction from this. 
https://github.com/ImageMagick/ImageMagick-Windows

and environment is like this OS - windows 10 64bit / compiler VS2015 professional with update3 / 
so many errors. a little i will show. 
i dont understand, how to solve problem. plz help me how to do. 

28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\xstddef(377): error C2338: The C++ Standard doesn't provide a hash for this type. (compiling source file ..\..\harfbuzz\src\hb-aat-layout.cc)
28>  d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\harfbuzz\src\hb-algs.hh(251): note: see reference to class template instantiation 'std::hash<hb_bit_page_t>' being compiled (compiling source file ..\..\harfbuzz\src\hb-aat-layout.cc)
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\xstddef(377): error C2338: The C++ Standard doesn't provide a hash for this type. (compiling source file ..\..\harfbuzz\src\hb-buffer.cc)
28>  d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\harfbuzz\src\hb-algs.hh(251): note: see reference to class template instantiation 'std::hash<hb_bit_page_t>' being compiled (compiling source file ..\..\harfbuzz\src\hb-buffer.cc)
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\xstddef(377): error C2338: The C++ Standard doesn't provide a hash for this type. (compiling source file ..\..\harfbuzz\src\hb-aat-map.cc)
28>  d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\harfbuzz\src\hb-algs.hh(251): note: see reference to class template instantiation 'std::hash<hb_bit_page_t>' being compiled (compiling source file ..\..\harfbuzz\src\hb-aat-map.cc)

**Comment by dlemstra** [comment]
Can you add the following to `harfbuzz\Config.txt`, run `configure.exe` again and then try to build again?

```
[VISUAL_STUDIO]
2017
```

**Comment by papafrog77** [comment, marked as answer]
thank you for your answer. 

is that meaning, i must use VS2017 ? instead of VS2015 . ?

**Reply by dlemstra** [reply]
Yes, this means that the build requires at least VS2017. Can you confirm that this fixes your build?

**Reply by papafrog77** [reply]
thank you for your help. 

I think this may have progress. 
but still got some error. 

i attached error file. but some language korean.. so just focus on error code, 

while compile CORE_MagickCore  project.
magickcore\color.c(1331): error C2220:
configure.c(482): error C2220
delegate.c(1386): error C2220:
log.c(566): error C2220
magic.c(511): error C2220:
mime.c(479): error C2220:
policy.c(415): error C2220: 

got errors so i can not make CORE_MagickCore lib so can not procede more. 

if you have some idea , plz help me. 
i also tested with GraphicsMagic, that's different with IM, GM works with VS2013  well. but i think ImageMagick is better so  i try this.

**Reply by dlemstra** [reply]
C2220 means you are getting a warning that is treated as an error. Can you translate the warning/error message or the error log so I can try to translate it myself?

**Reply by papafrog77** [reply]
37>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\configure.c(482): error C2220: 경고가 오류로 처리되어 생성된 'object' 파일이 없습니다.
37>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\configure.c(482): warning C4090: '함수': 'const' 한정자가 서로 다릅니다.

convert shortly.  
 error C2220: warning make error so can not make object file. 
 warning C4090 : function : "const" modifier is different.

**Reply by dlemstra** [reply]
We will probably need a proper patch for this to make it work with `VS2015` but what happens when you remove `[TREAT_WARNING_AS_ERROR]` from the `VisualMagick\MagickCore\config.txt` file?

**Reply by papafrog77** [reply]
Thank you for your reply. 
as you said, so i changed config.txt , and compile again, same error occur. 
and i changed compiler to VS2017. 

38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\color.c(1331): error C2220: 경고가 오류로 처리되어 생성된 'object' 파일이 없습니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\color.c(1331): warning C4090: '함수': 'const' 한정자가 서로 다릅니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\delegate.c(1386): error C2220: 경고가 오류로 처리되어 생성된 'object' 파일이 없습니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\delegate.c(1386): warning C4090: '함수': 'const' 한정자가 서로 다릅니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\log.c(566): error C2220: 경고가 오류로 처리되어 생성된 'object' 파일이 없습니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\log.c(566): warning C4090: '함수': 'const' 한정자가 서로 다릅니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\magic.c(511): error C2220: 경고가 오류로 처리되어 생성된 'object' 파일이 없습니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\magic.c(511): warning C4090: '함수': 'const' 한정자가 서로 다릅니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\mime.c(479): error C2220: 경고가 오류로 처리되어 생성된 'object' 파일이 없습니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\mime.c(479): warning C4090: '함수': 'const' 한정자가 서로 다릅니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\policy.c(415): error C2220: 경고가 오류로 처리되어 생성된 'object' 파일이 없습니다.
38>d:\zwork\dev\project - xcolor rip\image-lib-for-whiteink\imagemagick-src\imagemagick-7.1.0.52\imagemagick\magickcore\policy.c(415): warning C4090: '함수': 'const' 한정자가 서로 다릅니다.

all occur in Magickcore . 
and lib can not make so using this lib all error. of course.

**Reply by dlemstra** [reply]
Did you also run `configure.exe` after making that change?

**Reply by papafrog77** [reply]
Yes. of course.  same results.

**Reply by dlemstra** [reply]
Can you give it another try with the latest commit?
